src/smtp_server.c

```diff
@@ -122,58 +138,70 @@ int smtp_server_response(bfd_t *f, int code, const char *message)
 
 int smtp_server_process(struct smtp_server_context *ctx, const char *cmd, const char *arg, bfd_t *stream)
 {
-	int schs, continue_session = 1;
+	int disconnect = 1;
+	int hdlr_idx;
+	int code;
+
 	struct smtp_cmd_hdlr_list *hlink;
+	struct smtp_cmd_hdlr *cmd_hdlr;
+	char *message;
 
-	do {
-		/* Save ctx->node to local var, since it can be changed by a
-		 * command handler *while* we are walking the handler list */
-		struct smtp_cmd_tree *node = ctx->node;
+	code = 0;
+	message = NULL;
+	hdlr_idx = smtp_get_hdlr_idx(cmd);
 
-		ctx->code = 0;
-		ctx->message = ctx->prev_message = NULL;
+	if (hdlr_idx != -1) {
+		/* Get the structure with specific handler */
+		cmd_hdlr = &smtp_cmd_hdlrs[hdlr_idx];
 
-		list_for_each_entry(hlink, &node->hdlrs, lh) {
+		/* Call the handler */
+		disconnect = !cmd_hdlr->smtp_preprocess_hdlr(ctx, cmd, arg, stream);
```

Comment:
The 'disconnect' values are consistent with the way they are used back in __smtp_server_run() but inconsistent with the variable name (disconnect is 0 when the session must be disconnected and vice-versa).

For code clarity, I suggest switching values here in smtp_server_process() (initialize the variable to 0 and remove negation before assigning the return value from the handler) and negating the return value of smtp_server_process() back in __smtp_server_run().